File: server/src/main/java/gg/mineral/server/network/packet/handshake/serverbound/HandshakePacket.java
package gg.mineral.server.network.packet.handshake.serverbound;

import gg.mineral.api.network.connection.Connection;
import gg.mineral.api.network.packet.Packet;
import gg.mineral.server.network.connection.ConnectionImpl;
import gg.mineral.server.network.protocol.ProtocolState;
import io.netty.buffer.ByteBuf;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Accessors(fluent = true)
public final class HandshakePacket implements Packet.INCOMING {
    private int protocol, port, nextState;
    private String serverAddress;

    @Override
    public void received(Connection connection) {
        if (connection instanceof ConnectionImpl impl) {
            impl.setProtocolVersion((byte) protocol);
            impl.setProtocolState(ProtocolState.getState(nextState));
        }
    }

    @Override
    public void deserialize(ByteBuf is) {
        protocol = readVarInt(is);
        serverAddress = readString(is);
        port = is.readUnsignedShort();
        nextState = readVarInt(is);
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/handshake/serverbound/HandshakePacket.kt
package gg.mineral.server.network.packet.handshake.serverbound

import gg.mineral.api.network.connection.Connection
import gg.mineral.api.network.packet.Packet
import gg.mineral.server.network.connection.ConnectionImpl
import gg.mineral.server.network.protocol.ProtocolState
import io.netty.buffer.ByteBuf

class HandshakePacket(
    var protocol: Int = 0,
    var port: Int = 0,
    var nextState: Int = 0,
    var serverAddress: String? = null
) : Packet.Incoming, Packet.EventLoopHandler {
    override fun receivedEventLoop(connection: Connection) {
        if (connection is ConnectionImpl) {
            connection.protocolVersion = protocol.toByte()
            connection.protocolState = ProtocolState.getState(nextState)
        }
    }

    override fun deserialize(`is`: ByteBuf) {
        protocol = `is`.readVarInt()
        serverAddress = `is`.readString()
        port = `is`.readUnsignedShort()
        nextState = `is`.readVarInt()
    }
}
